libsolidity/analysis/TypeChecker.cpp

```diff
@@ -317,6 +306,51 @@ void TypeChecker::checkContractIllegalOverrides(ContractDefinition const& _contr
 	}
 }
 
+void TypeChecker::checkFunctionOverride(FunctionDefinition const& function, FunctionDefinition const& super)
+{
+	FunctionType functionType(function);
+	FunctionType superType(super);
+
+	if (!functionType.hasEqualArgumentTypes(superType))
+		return;
+
+	if (function.visibility() != super.visibility())
+		m_errorReporter.typeError(
+			function.location(),
+			"Overriding function visibility differs from " + super.fullyQualifiedName() + "."
+		);
+
+	if (function.isDeclaredConst() && !super.isDeclaredConst())
+		m_errorReporter.typeError(
+			function.location(),
+			"Overriding function should not be declared constant."
+		);
+
+	if (!function.isDeclaredConst() && super.isDeclaredConst())
+		m_errorReporter.typeError(
+			function.location(),
+			"Overriding function should be declared constant."
+		);
+
+	if (function.isPayable() && !super.isPayable())
+		m_errorReporter.typeError(
+			function.location(),
+			"Overriding function should not be declared payable."
+		);
+
+	if (!function.isPayable() && super.isPayable())
+		m_errorReporter.typeError(
+			function.location(),
+			"Overriding function should be declared payable."
+		);
+
+	if (functionType != superType)
+		m_errorReporter.typeError(
+			function.location(),
+			"Overriding function return types differ from " + super.fullyQualifiedName() + "."
```

Comment:
Good idea, I wasn't aware of SecondarySourceLocationn. I have updated the PR accordingly.